src/replication.c

```diff
@@ -704,7 +704,14 @@ void replicationFeedMonitors(client *c, list *monitors, int dictid, robj **argv,
     listRewind(monitors, &li);
     while ((ln = listNext(&li))) {
         client *monitor = ln->value;
-        addReply(monitor, cmdobj);
+        if (monitor->resp > 2) {
+            monitor->flag.pushing = 1;
```

Comment:
we should undo the pushing flag in the last?

Reply:
```
Logged crash report (pid 67673):
=== VALKEY BUG REPORT START: Cut & paste starting from here ===
67673:M 16 Dec 2024 12:11:36.111 # === ASSERTION FAILED CLIENT CONTEXT ===
67673:M 16 Dec 2024 12:11:36.111 # client->flags = 18014398509481990
67673:M 16 Dec 2024 12:11:36.111 # client->conn = fd=13
67673:M 16 Dec 2024 12:11:36.111 # client->argc = 0
67673:M 16 Dec 2024 12:11:36.111 # === RECURSIVE ASSERTION FAILED ===
67673:M 16 Dec 2024 12:11:36.111 # ==> networking.c:1087 'c->flag.pushing' is not true
```
without this flag set instance is crashing - I just saw its used everywhere when push i used 
```
void addReplyPubsubMessage(client *c, robj *channel, robj *msg, robj *message_bulk) {
    struct ClientFlags old_flags = c->flag;
    c->flag.pushing = 1;
    if (c->resp == 2)
        addReply(c, shared.mbulkhdr[3]);
```
I did the same.

Reply:
@enjoy-binbin Well spotted! Yes, we must reset it, right? Otherwise there will be some problem if the client sends commands. If the pushing flag is set, the replies are queued and not sent to the client until later.

Btw, @KowalczykBartek you didn't add a test case where the client sends a PING? We should add it and the test case needs to check that the pong-reply and monitored-ping push come in the right order. The pong reply should come first.

This is the code in networking.c that cares about this flag:

```C
    /* If we're processing a push message into the current client (i.e. executing PUBLISH
     * to a channel which we are subscribed to, then we wanna postpone that message to be added
     * after the command's reply (specifically important during multi-exec). the exception is
     * the SUBSCRIBE command family, which (currently) have a push message instead of a proper reply.
     * The check for executing_client also avoids affecting push messages that are part of eviction.
     * Check CLIENT_PUSHING first to avoid race conditions, as it's absent in module's fake client. */
    if (c->flag.pushing && c == server.current_client && server.executing_client &&
        !cmdHasPushAsReply(server.executing_client->cmd)) {
        _addReplyProtoToList(c, server.pending_push_messages, s, len);
        return;
    }
```

Reply:
@KowalczykBartek It seem you misunderstood the comment. We should *reset the flag to its previous value after pushing*.

```
int old_pushing = c->flags.pushing;
c->flags.pushing = 1;
push stuff...
c->flags.pushing = old_pushing;
```

Reply:
I pushed both if statement for flag reset and test but I faced some issue with tcl, that's what valkey returns (I see this with my toy client) for combination of monitor + ping
```
+PONG
>2
$7
monitor
+1734353183.716803 [0 127.0.0.1:50339] "PING"
```
but client in test doesn't see this PONG :/ 

```tlc

    test {MONITOR should came after PONG reply} {
        set rd [valkey_deferring_client]
        $rd HELLO 3
        $rd read ; # Consume the HELLO reply

        $rd monitor
        $rd read ; # Consume the MONITOR reply
        $rd readraw 1;

        r ping

        assert_equal "+pong" [$rd read]
        assert_equal ">2" [$rd read]
        assert_equal "\$7" [$rd read]
        assert_equal "monitor" [$rd read]
        assert_match {*"ping"*} [$rd read]

        $rd close
    }
```
in this line 
```
 assert_equal "+pong" [$rd read]
```
it see `>2`, so, ping is skipped because of some reason, do you have any idea what is going on ?

Reply:
`r` is one client and `$rd` is another. You can see the PONG as the return value of `r ping`:

    assert_equal PONG [r ping]

... but what we need to test here is to send the command from the monitoring client, i.e. using `$rd ping`.

Reply:
yeah.. I am blind, sorry for this :/ I pushed fixed test case